src/mesh/RTree.hpp

```diff
@@ -1,21 +1,161 @@
+  static constexpr Primitive value = Primitive::Vertex;
+};
+template <>
+struct as_primitive_enum<mesh::Edge> {
+  static constexpr Primitive value = Primitive::Edge;
+};
+template <>
+struct as_primitive_enum<mesh::Triangle> {
+  static constexpr Primitive value = Primitive::Triangle;
+};
+template <>
+struct as_primitive_enum<mesh::Quad> {
+  static constexpr Primitive value = Primitive::Quad;
+};
+
+/** Binds an Index and Primitive into a type
+ * \see AABBGenerator
+ * \see indexPrimitives
+ */
+struct PrimitiveIndex {
+  Primitive type;
+  size_t    index;
+};
+
+/// Standard equality test for PrimitiveIndex
+bool operator==(const PrimitiveIndex& lhs, const PrimitiveIndex& rhs);
+
+/// Standard non-equality test for PrimitiveIndex
+bool operator!=(const PrimitiveIndex& lhs, const PrimitiveIndex& rhs);
+
+/// A standard print operator for PrimitiveIndex
+std::ostream& operator<<(std::ostream& out, PrimitiveIndex val);
+
+/// The axis aligned bounding box based on the Vertex Type
+using AABB = boost::geometry::model::box<Eigen::VectorXd>;
+
+/** Holding a reference to a Mesh it is a functor for packing
+ *  PrimitiveIndex into AABBs.
+ */
+class AABBGenerator
+{
+public:
+  /// Constructs a generator for a given mesh
+  explicit AABBGenerator(const mesh::Mesh &mesh)
+      : mesh_(mesh){};
+
+  /** constructs a AABB of a PrimitiveIndex.
+   *
+   * This operator takes a PrimitiveIndex and fetches the actual primitive from the mesh.
+   * It then constructs an AABB using boost::geometry::return_envelope() and returns it.
+   *
+   * \param pi the index to build the AABB for
+   * \returns a AABB for the primitive
+   */
+  AABB operator()(const PrimitiveIndex &pi) const
+  {
+    switch (pi.type) {
+    case (Primitive::Vertex):
+      return boost::geometry::return_envelope<AABB>(mesh_.vertices()[pi.index]);
+    case (Primitive::Edge):
+      return boost::geometry::return_envelope<AABB>(mesh_.edges()[pi.index]);
+    case (Primitive::Triangle):
+      return boost::geometry::return_envelope<AABB>(mesh_.triangles()[pi.index]);
+    case (Primitive::Quad):
+      return boost::geometry::return_envelope<AABB>(mesh_.quads()[pi.index]);
+    default:
+      assertion(false, "AABB generation for this Primitive is not implemented!")
+    }
+  }
+
+private:
+  /// The mesh used look the primitives up
+  const mesh::Mesh &mesh_;
+};
+
+/// The rtree capable of indexing primitives of an entire Mesh
+using PrimitiveRTree = boost::geometry::index::rtree<std::pair<AABB, PrimitiveIndex>, boost::geometry::index::rstar<16>>;
+
+/// The shared_ptr convenience type for PrimitiveRTree
+using PtrPrimitiveRTree = std::shared_ptr<PrimitiveRTree>;
+
+/** indexes a container of primitives into an rtree.
+ *
+ * The algorithm indexes every primitive of the given container using the passed generator.
+ * It inserts its result and the PrimitiveIndex into the rtree as a std::pair.
+ *
+ * \param[inout] rtree the rtree to index into
```

Comment:
`\param[in,out]`

Reply:
Thanks, for little typos like this you could try the new suggestion feature.